### ISSUE TITLE
a request

### ISSUE THREAD

**Issue by COST-97** [cics-nd/ar-pde-cnn#3]
Hello! In your paper, formula (19) is used to design the loss function of the model, and the corresponding code is in `bayesNN.py.` 
But I can't understand the implementation principle in the `calc_neg_log_joint` function. 
Could you explain the principle of formula (19) implemented in the program further?
Thank you very much!

**Comment by NickGeneva** [comment]
Hi,

Good question. When dealing with the optimization of density functions (particularly Gaussian which are fairly common) the optimization objective is posed as minimizing the negative log of the target distribution. To my understanding, this is typically done (at least in this case) because this simplifies the calculations needed and results in many normalizing constants not being needed to be computed for optimization of the hyper-parameters. In other words we only need to calculate parts of the distribution parameters we are trying to optimize exist.

So in this case three components:
1) The log of the Gaussian likelihood function of the output.
`log_likelihood = ntrain / output.size(0) * (- 0.5 * self.model.log_beta.exp()
                            * (target - output).pow(2).sum()
                            + 0.5 * target.numel() * self.model.log_beta)`

2) The log of the prior on the weights of the NN (model paramters)  which due to out choice is a Student T distribution which is why this looks a little different than what is in the paper. We pre-combine the Normal(w|0,\alpha^{1}I)Gamma(\alpha|a,b).
`for param in self.model.features.parameters():`
&nbsp;&nbsp;&nbsp;&nbsp;`log_prob_prior_w += torch.log1p(0.5 / self.w_prior_rate * param.pow(2)).sum()`
        `log_prob_prior_w *= -(self.w_prior_shape + 0.5)`

3) And finally the log of the prior Gamma distribution on the learnable noise, beta, of the model.
`prior_log_beta = ((self.beta_prior_shape - 1.0) * self.model.log_beta
                    - self.model.log_beta.exp() * self.beta_prior_rate)`

So for example, in three when looking at the true [Gamma distribution](https://en.wikipedia.org/wiki/Gamma_distribution) you'll see that even after taking the log we have ditched a large number of its normalizing coefficients. This is because we are only concerned with the optimization of beta (so x on wikipedia) thus all parts of the distribution that do not contain it play no role in the gradient and don't need to be calculated.

We return the negative sum of all these components which is minimized. Note that the optimization of this alone results in a MAP estimate which is analogous to a regularized maximum likelihood. Sampling from posterior is the tricky part which is where Bayesian deep learning approximations typically come into play.

**Comment by COST-97** [comment]
Hello! May I have a look at the mathematical expressions corresponding to these three values in the code?
 After I substituted the Gaussian distribution and the gamma distribution into formula (19), it was slightly different from the code implementation, so that I could not understand some details in the code. 
Specifically, `log_likelihood = ntrain / output.size (0) * (-0.5 * self.model.log_beta.exp () * (target-output) .pow (2) .sum () + 0.5 * in the first part target.numel () * self.model.log_beta)` ,why multiply by `ntrain / output.size (0)`.
The second part of `log_prob_prior_w + = torch.log1p (0.5 / self.w_prior_rate * param.pow (2)). sum ()` Why is `torch.log1p`.
`log_prob_prior_w * =-(self.w_prior_shape + 0.5)` Why multiply `(self.w_prior_shape + 0.5)`.
In the third part, `priority_log_beta = ((self.beta_prior_shape-1.0) * self.model.log_beta-self.model.log_beta.exp () * self.beta_prior_rate)` Why multiply `self.beta_prior_rate`.
thank you very much.

**Comment by NickGeneva** [comment]
See responses to each part below:

1) Because when evaluating the likelihood we originally propose it with respect to the entire training data set. However, here we are using mini-batches (an approximation) thus we must scale the likelihood to what it would be if we were evaluating it for the entire training data-set in a single batch. Hence, we multiply by `ntrain / output.size (0)`.

2) This is due to the form of the Student T distribution and how the hyper-parameters from the gamma distribution come into the play. The log1p(x) = log(1+x) which is clearly kernel of the [Students T pdf](https://wikimedia.org/api/rest_v1/media/math/render/svg/e293ade2e368fb5618602ddc310ef149b88c1496). The rest I won't go into too much detail but if you seek further information see Eq. 2.159 in Pattern Recognition and Machine Learning by Bishop to better understand the marginalization.

3)  If you looked at the Gamma distribution I linked before ([top equation here](https://wikimedia.org/api/rest_v1/media/math/render/svg/9341808f17a3081746f5416921a95561e8bc81bf)) the `self.model.log_beta.exp () * self.beta_prior_rate` is from the e^{-\beta x} component.

**Comment by COST-97** [comment]
I think about it again.
Thank you very much for your explanation.
